[Bharath] iP

bharathcs opened the pull request:
# A Pull Request

This application is based off [SE-Edu's](https://se-education.org/) _Duke_. 
It ~~lets you lie to yourself that you're organised~~ keeps you on track for deadlines. 

**What Duke could do before this:**

- Say Hi

**What Duke can do now:**

1. Add tasks (deadlines, todos, and events too)
2. Handle dates
3. List, delete, find tasks
4. More coming up.

Give it a go 🚀 

```java
Duke dk = new Duke();
dk.run();

/*
Hello from
 ____        _        
|  _ \ _   _| | _____ 
| | | | | | | |/ / _ \
| |_| | |_| |   <  __/
|____/ \__,_|_|\_\___|
*/
```

## Fuss free install

Download the fat JAR from [here](https://github.com/bharathcs/cs2103t-ip/releases/tag/v0.1) and run `java -jar path/to/downloaded.jar` in your terminal. For max java, enjoy with a good cup of coffee.

## Wishlist

- [x] List, delete, find tasks
- [x] Error handling to ward off bad user inputs
- [x] Task types - Todo, Deadline, Event
- [ ] Functional Command class
- [ ] GUI

> “If debugging is the process of removing software bugs, then programming must be the process of putting them in.”
> ― Edsger W. Dijkstra

Commit:
Add Gradle support

bharathcs committed:
Amend gitignore

bharathcs committed:
Add I/O skeleton to Duke.

bharathcs committed:
Modify intro text

bharathcs committed:
Refactor behaviour logic out to DukeLogic class.

bharathcs committed:
Handle adding and listing tasks

- also important, empty lines no longer cause a task to be created.

bharathcs committed:
Appease prettier & checkstyle.

bharathcs committed:
Add documentation for public functions and classes.

bharathcs committed:
Implement done command for tasks

- Include new field for Tasks ( `isComplete` )
- Add command handling for done
- add padding for integers to ensure consistency in output.

bharathcs committed:
Add task types to handle TODO, DEADLINE, EVENTs.

bharathcs committed:
Add error handling for badly written commands.

bharathcs committed:
Fix output for adding tasks and listing

bharathcs committed:
Add in basic automated I/O testing.

bharathcs committed:
Improve exception handling from earlier.

bharathcs committed:
Modify google checkstyle to support Prettier-Java plugin config.

bharathcs committed:
Add deletion abilities and use enums for individual task actions.

bharathcs committed:
Improve string checking (drop the unnecessary higher order fns)

bharathcs committed:
Add in testing files (missed out from bef)

bharathcs commented:
3b7463b (HEAD -> master, origin/master, origin/HEAD) Add in testing files (missed out from bef)
6f074ee Improve string checking (drop the unnecessary higher order fns)
30cdf3a (tag: Level-6, tag: A-Enums, tag: A-Collections) Add deletion abilities and use enums for individual task actions.
92c8fe7 Modify google checkstyle to support Prettier-Java plugin config.
e79db8a (tag: Level-5, tag: A-Exceptions) Improve exception handling from earlier.
741e759 (tag: A-TextUiTesting) Add in basic automated I/O testing.
15f09f4 (tag: Level-4, tag: A-Inheritance) Fix output for adding tasks and listing
49ce696 Add error handling for badly written commands.
679b4cc Add task types to handle TODO, DEADLINE, EVENTs.
3b0a55b (tag: Level-3, tag: A-Classes) Implement done command for tasks
53ba2f4 (tag: Level-2) Add documentation for public functions and classes.
cac8d27 Handle adding and listing tasks
89529ae Refactor behaviour logic out to DukeLogic class.
c326726 (tag: Level-1) Appease prettier & checkstyle.
7736664 Modify intro text
f577d7e Add I/O skeleton to Duke.
f38971f Amend gitignore

bharathcs committed:
Add functions in Task to convert to/from backup.

bharathcs committed:
Add backup/restore to database to Duke.

bharathcs committed:
Add documentation for the new public functions.

bharathcs committed:
Add useful regex to DateParser.

bharathcs committed:
Use proper Date types in Tasks.

bharathcs committed:
Update database functions to support Date type.,

bharathcs committed:
gitignore /data folder wherever found.

bharathcs committed:
Add isComplete to internal string and task representations.

bharathcs committed:
Merge branch 'branch-Level-7'

bharathcs committed:
Merge branch 'branch-Level-8'

bharathcs committed:
Update test results for Level-7 and 8.

Removed EXPECTED-UNIX as it is is overwritten by values in expected anyway.

bharathcs committed:
mv `DukeStorage` -> `Storage`.

bharathcs committed:
Go back to original google-checkstyle rules.

bharathcs committed:
Change indent rules to CS2103T standards.

bharathcs committed:
Add task list functions to Storage.

bharathcs committed:
Refactor into packages and update runtest executable.

bharathcs committed:
Do up the OOP format with non-static classes.

bharathcs committed:
Reword tasksList -> tasks for consistency.

bharathcs committed:
Add JUnit testing.

bharathcs committed:
Appease coding standards.

bharathcs committed:
Add in missing JavaDoc

bharathcs committed:
Add in find abilities

bharathcs committed:
Merge branch 'branch-A-JavaDoc'

bharathcs committed:
Merge branch 'branch-A-CodingStandards'

bharathcs committed:
Merge branch 'branch-Level-9'

trxe commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,48 @@
+     */
+    public static void main(String[] args) {
+        new Duke().run();
+        System.out.println("a0");
```
(line 46) just curious, what's this for

bharathcs replied:
Oh haha oops! It was supposed to be removed, it was a debugging statement that I accidentally left in. I was testing out running `.run()` multiple times and deleted the extra calls but forgot this line. Thanks!

trxe commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,101 @@
+        StringBuilder save = new StringBuilder();
+
+        taskList
+                .stream()
```
Nitpicking here but I think the line break here is probably unnecessary

trxe commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,175 @@
+     *
+     * @param isComplete new completion status
+     */
+    public void markComplete(boolean isComplete) {
```
Usually it would be called `setComplete`, but honestly this is already quite clear too

trxe commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,48 @@
+ */
+public class Duke {
+
+    private final Ui ui;
```
I'm not too sure myself if **all** `final` variables have to be in caps (doesn't make sense to me either if it's not `static`) but for the record it is stated [here](https://se-education.org/guides/conventions/java/intermediate.html#naming) to be the case

trxe reviewed (commented):
Your code is so clean I'm envious. It was hard to find any violations to comment on. Good on you man, thanks for your work

bharathcs committed:
Use Gradle 6.3 to support JDK 14.

bharathcs committed:
Add in AB3 checkstyle and suppression.xml for gradle

bharathcs committed:
Appease checkstyle and heed reviewer suggestions.

bharathcs commented:
Thanks @trxe !

bharathcs committed:
Remove debug statement.

bharathcs committed:
Add JavaFX to Gradle.

elroygohjy commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,177 @@
+            }
+            LocalDateTime startDate;
+            LocalDateTime endDate;
+            try {
+                String[] arr = args[1].split(" - ");
+                if (arr.length != 2) {
+                    DateParser.parseDateTimeInput(""); // throws error
+                }
+                startDate = DateParser.parseDateTimeInput(arr[0]);
+                endDate = DateParser.parseDateTimeInput(arr[1]);
+            } catch (DateTimeParseException e) {
+                throw new InvalidTaskException(
+                    "The two dates for deadline creation could not be parsed. Expected:\n"
+                        + "2 dates separated by ' - '. Dates come in the forms: "
+                        + "'YYYY-MM-DD' or 'YYYY-MM-DD HHMM' (Time in 24hr format)."
+                );
+            }
```
Do you think you can extract this part out as a method to simplify the code?

elroygohjy commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,177 @@
+            }
+            return new EventTask(args[0].trim(), startDate, endDate);
+        case DEADLINE:
+            args = inputString.split(" /by ");
+            if (args.length != 2) {
+                throw new InvalidTaskException("Expected '{title} /by {dates}' for deadline tasks");
+            }
+            LocalDateTime date;
+            try {
+                date = DateParser.parseDateTimeInput(args[1].trim());
+            } catch (DateTimeParseException e) {
+                throw new InvalidTaskException(
+                    "Date for event creation could not be parsed. Expected:\n"
+                        + "'YYYY-MM-DD' or 'YYYY-MM-DD HHMM' (Time in 24hr format)."
+                );
+            }
```
Similarly for this, do you think that you can extract it out as a method? Perhaps what you can do is to catch both exceptions for event and deadline together after you extract them as methods?

elroygohjy commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,99 @@
+
+        ArrayList<Task> taskArr = new ArrayList<>();
+        try {
+            Stream<String> databaseToString = Files.lines(filePath);
+            databaseToString.forEach(
+                taskString -> {
+                    try {
+                        Task task = Task.stringToTask(taskString);
+                        taskArr.add(task);
+                    } catch (IllegalArgumentException e) {
+                        failedParses.add(taskString);
+                    }
+                }
+            );
+            databaseToString.close();
```
Do you think this part can extract out as a method?

bharathcs replied:
Yep, makes sense. Thanks!

elroygohjy reviewed (commented):
I think you have done a great job! You just need to touch up on the parts I mentioned

Tetrerox commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,139 @@
+package duke;
+
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import duke.tasks.InvalidTaskException;
+import duke.tasks.Task;
+import duke.tasks.TaskList;
```
I think that you did a good job of organising the imports! 👍

Tetrerox commented on src/main/java/duke/tasks/DateParser.java:
```diff
@@ -0,0 +1,78 @@
+    public static String toDatabaseFormat(LocalDateTime dateTime) {
+        return dateTime.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm"));
+    }
+}
```
I like the names of these methods! It makes the code very easy to understand and follow.

Tetrerox commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,177 @@
+     */
+    public static Task stringToTask(String stringifiedTask) {
+        // {TYPE}|{DESCRIPTION}|{DATE or DATES or BLANK}
+        String[] taskAttr = stringifiedTask.split(Pattern.quote(DELIMITER));
```
Maybe the variable name could be made more understandable?

Tetrerox commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,72 @@
+     *
+     * @param index index of the task to return
+     * @return the element at the specified position in this list
+     * @throws IndexOutOfBoundsException if there is no such index in list.
```
I believe that you do not have to include this line as your method does not explicitly state that it throws something. Not sure if you have to include it in because of the .get method.

bharathcs replied:
Hmm yes, the `get` method throws it, I copied it from its javadocs, but you're right, I should look into this.

Tetrerox commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,72 @@
+     * Removes a specific task from internal list. List will be empty after this call.
+     *
+     * @param index index of the task to return
+     * @throws IndexOutOfBoundsException if there is no such index in list.
```
Same thing over here

Tetrerox reviewed (commented):
I found it extremely difficult to find any coding standards violations. Keep up the good work 👍

bharathcs committed:
Listen to more PR suggestions.

bharathcs committed:
Follow SE-EDU steps for JavaFX Steps 1 - 3.

bharathcs committed:
Undo SE-EDU steps 1-3

bharathcs committed:
Follow SE-EDU steps for JavaFX Step 4.

bharathcs committed:
Customise GUI and fix minor output bug.

bharathcs committed:
Merge branch 'branch-Level-10'

bharathcs committed:
Remove unnecessary legacy code (from CLI era).

bharathcs committed:
Improve comprehension (comments, functions, vars).

bharathcs committed:
Add assert

bharathcs committed:
Appease checkstyle

bharathcs committed:
Fix tests

bharathcs committed:
Create .gitkeep

bharathcs committed:
Create new Github Action Workflows.

Checkstyle and Tests are not automatically checked on PRs and pushes.
Fat JARs have to be manually built and uploaded to each version release.

Repetitive manual actions like these waste developer's time.

Create Github action workflows to automatically have Gradle run the
`check` task to verify PRs and pushes to master maintain repo quality.
Then the `shadowJar` task is used to create the fat JAR and then upload
to the `latest` release.

Gradle and shadowJar already has well implemented ways to test and
create fat JARs and using Github's action process automates this.

Note that for shadowJar, it runs on version releases, which are
identified as a tag named 'v*' (e.g. v1.0).

bharathcs committed:
Add C-Sort functionality.

Tasks are presented in the order they are added.

User will likely want to see urgent deadlines and events before other
tasks.

Add an abstract function to `Task` to obtain a numeric value of urgency.
Then provide a comparator in `Task` for use in `Parser` to sort by
urgency before presenting users with a list.

To show this view, Duke must reorder the tasks in terms of the date
associated with it. Therefore, the different Task subclasses must
provide a numeric value to compare dates against each other.
Converting them to Unix timestamp results in an easily handled number.
This also lets Todo tasks (without any dates) to occupy the bottom with
`Long.MAX` as it has no date (i.e. no urgency).

In the future, any comparators can involve other attributes like
priority in its calculation to create a final comparison.

bharathcs committed:
Fix filepath to use a .txt and save regularly.

bharathcs committed:
Add other assertions.

bharathcs committed:
Merge pull request #2 from bharathcs/branch-A-Assertions

Add assertions to codebase.

bharathcs committed:
Merge branch 'master' into branch-A-CodeQuality

bharathcs committed:
Merge pull request #3 from bharathcs/branch-A-CodeQuality

Improve code quality in repo.

bharathcs committed:
Merge branch 'master' into branch-C-Sort

bharathcs committed:
Minor changes to appease checkstyle.

bharathcs committed:
Standardise form of AM / PM.

Github machines appear to output AM instead of am like local dev machine. Inconsistency could not be solved with code, appears to be a Locale setting based on the JVM.

Fails check test

Fix with a `String.replace(..)` to force consistency across every platform.

bharathcs committed:
Merge pull request #4 from bharathcs/branch-C-Sort

Add C-Sort functionality.

bharathcs committed:
Remove ./text-ui-test and extra checkstyle.

These are unnecessary now

bharathcs committed:
Add user guide for Duke.

bharathcs committed:
Set theme jekyll-theme-cayman

bharathcs committed:
Try changing image size for gh-pages

bharathcs committed:
Add more GUI tweaks

Rounded corners, drop shadows, margins and new profile pictures.

bharathcs committed:
Fix build.gradle bug causing unpacked dependencies in fat JAR

bharathcs committed:
Fix build.gradle bug causing unpacked dependencies in fat JAR

bharathcs committed:
Disable user input and shutdown after "bye" command.

After a time of 1.6 seconds.

bharathcs committed:
Merge branch 'branch-A-BetterGui'

bharathcs committed:
Update README.md

Clean up newlines

bharathcs committed:
Avoid useless gradle checks on non-code files.

bharathcs committed:
Add greeting at the start

bharathcs committed:
Set theme jekyll-theme-cayman

bharathcs committed:
Update README.md

bharathcs committed:
Fix spelling.
